File: dast/helpers/filesystem-snapshot.ts
/**
 * REDACTS DAST — Filesystem Snapshot Helper
 * ==========================================
 * Takes before/after snapshots of the REDCap filesystem to detect
 * unexpected file changes during operations (especially upgrades).
 * Compares file lists, sizes, and content hashes.
 */

import { execSync } from "child_process";
import * as fs from "fs";
import * as path from "path";
import * as crypto from "crypto";

export interface FileEntry {
  path: string;
  size: number;
  hash: string;
  mtime: string;
}

export interface SnapshotDiff {
  added: string[];
  removed: string[];
  modified: string[];
  unchanged: number;
  suspiciousFiles: SuspiciousFile[];
}

export interface SuspiciousFile {
  path: string;
  reason: string;
  detail: string;
}

/**
 * Take a snapshot of all files under a given directory.
 */
export function takeSnapshot(dir: string): Map<string, FileEntry> {
  const entries = new Map<string, FileEntry>();
  if (!fs.existsSync(dir)) return entries;

  walkDir(dir, dir, entries);
  return entries;
}

function walkDir(
  root: string,
  current: string,
  entries: Map<string, FileEntry>
): void {
  for (const item of fs.readdirSync(current, { withFileTypes: true })) {
    const full = path.join(current, item.name);
    const rel = path.relative(root, full);

    if (item.isDirectory()) {
      walkDir(root, full, entries);
    } else if (item.isFile()) {
      const stat = fs.statSync(full);
      const content = fs.readFileSync(full);
      const hash = crypto.createHash("sha256").update(content).digest("hex");

      entries.set(rel, {
        path: rel,
        size: stat.size,
        hash,
        mtime: stat.mtime.toISOString(),
      });
    }
  }
}

/**
 * Compare two snapshots and return the diff.
 */
export function diffSnapshots(
  before: Map<string, FileEntry>,
  after: Map<string, FileEntry>
): SnapshotDiff {
  const added: string[] = [];
  const removed: string[] = [];
  const modified: string[] = [];
  let unchanged = 0;
  const suspiciousFiles: SuspiciousFile[] = [];

  // Check for added and modified
  for (const [relPath, afterEntry] of after) {
    const beforeEntry = before.get(relPath);
    if (!beforeEntry) {
      added.push(relPath);
      checkSuspicious(relPath, afterEntry, "added", suspiciousFiles);
    } else if (beforeEntry.hash !== afterEntry.hash) {
      modified.push(relPath);
      checkSuspicious(relPath, afterEntry, "modified", suspiciousFiles);
    } else {
      unchanged++;
    }
  }

  // Check for removed
  for (const relPath of before.keys()) {
    if (!after.has(relPath)) {
      removed.push(relPath);
    }
  }

  return { added, removed, modified, unchanged, suspiciousFiles };
}

/**
 * Flag files that match INFINITERED/backdoor patterns.
 */
function checkSuspicious(
  relPath: string,
  entry: FileEntry,
  changeType: string,
  findings: SuspiciousFile[]
): void {
  const lc = relPath.toLowerCase();

  // Unexpected PHP in non-PHP directories
  if (lc.endsWith(".php") && (lc.includes("edocs/") || lc.includes("temp/"))) {
    findings.push({
      path: relPath,
      reason: `PHP file ${changeType} in data directory`,
      detail: `Size: ${entry.size}, Hash: ${entry.hash}`,
    });
  }

  // Hidden files
  if (path.basename(relPath).startsWith(".") && lc.endsWith(".php")) {
    findings.push({
      path: relPath,
      reason: `Hidden PHP file ${changeType}`,
      detail: `Size: ${entry.size}`,
    });
  }

  // Suspiciously large single-line PHP (obfuscated backdoors)
  if (lc.endsWith(".php") && entry.size > 50_000) {
    findings.push({
      path: relPath,
      reason: `Large PHP file ${changeType} (possible obfuscation)`,
      detail: `Size: ${entry.size} bytes`,
    });
  }

  // Files with encoded-looking names
  if (/[a-f0-9]{32,}/i.test(path.basename(relPath))) {
    findings.push({
      path: relPath,
      reason: `Hash-like filename ${changeType} (possible dropper)`,
      detail: relPath,
    });
  }
}

/**
 * Save snapshot diff to JSON for REDACTS reporting.
 */
export function saveDiffReport(
  diff: SnapshotDiff,
  outputPath: string
): void {
  fs.writeFileSync(outputPath, JSON.stringify(diff, null, 2), "utf-8");
}

File: dynamic/helpers/filesystem_snapshot.py
"""Filesystem snapshot helper - before/after diffs across REDCap operations.

Compares file lists, sizes, and content hashes so an upgrade or admin
action that quietly mutates the webroot does not slip past review.
"""

from __future__ import annotations

import hashlib
import json
import re
from dataclasses import dataclass, field
from pathlib import Path


@dataclass
class FileEntry:
    path: str
    size: int
    hash: str
    mtime: str


@dataclass
class SuspiciousFile:
    path: str
    reason: str
    detail: str


@dataclass
class SnapshotDiff:
    added: list[str] = field(default_factory=list)
    removed: list[str] = field(default_factory=list)
    modified: list[str] = field(default_factory=list)
    unchanged: int = 0
    suspicious_files: list[SuspiciousFile] = field(default_factory=list)


def take_snapshot(directory: str) -> dict[str, FileEntry]:
    """Take a snapshot of all files under a given directory."""
    root = Path(directory)
    entries: dict[str, FileEntry] = {}
    if not root.exists():
        return entries

    for item in root.rglob("*"):
        if not item.is_file():
            continue
        rel = str(item.relative_to(root))
        content = item.read_bytes()
        file_hash = hashlib.sha256(content).hexdigest()
        stat = item.stat()
        entries[rel] = FileEntry(
            path=rel,
            size=stat.st_size,
            hash=file_hash,
            mtime=str(stat.st_mtime),
        )

    return entries


def diff_snapshots(
    before: dict[str, FileEntry],
    after: dict[str, FileEntry],
) -> SnapshotDiff:
    """Compare two snapshots and return the diff."""
    diff = SnapshotDiff()

    # Check for added and modified
    for rel_path, after_entry in after.items():
        before_entry = before.get(rel_path)
        if before_entry is None:
            diff.added.append(rel_path)
            _check_suspicious(rel_path, after_entry, "added", diff.suspicious_files)
        elif before_entry.hash != after_entry.hash:
            diff.modified.append(rel_path)
            _check_suspicious(rel_path, after_entry, "modified", diff.suspicious_files)
        else:
            diff.unchanged += 1

    # Check for removed
    for rel_path in before:
        if rel_path not in after:
            diff.removed.append(rel_path)

    return diff


def _check_suspicious(
    rel_path: str,
    entry: FileEntry,
    change_type: str,
    findings: list[SuspiciousFile],
) -> None:
    """Flag files that match INFINITERED/backdoor patterns."""
    lc = rel_path.lower()
    basename = Path(rel_path).name

    # Unexpected PHP in non-PHP directories
    if lc.endswith(".php") and ("edocs/" in lc or "temp/" in lc):
        findings.append(SuspiciousFile(
            path=rel_path,
            reason=f"PHP file {change_type} in data directory",
            detail=f"Size: {entry.size}, Hash: {entry.hash}",
        ))

    # Hidden files
    if basename.startswith(".") and lc.endswith(".php"):
        findings.append(SuspiciousFile(
            path=rel_path,
            reason=f"Hidden PHP file {change_type}",
            detail=f"Size: {entry.size}",
        ))

    # Suspiciously large single-line PHP (obfuscated backdoors)
    if lc.endswith(".php") and entry.size > 50_000:
        findings.append(SuspiciousFile(
            path=rel_path,
            reason=f"Large PHP file {change_type} (possible obfuscation)",
            detail=f"Size: {entry.size} bytes",
        ))

    # Files with encoded-looking names
    if re.search(r"[a-f0-9]{32,}", basename, re.IGNORECASE):
        findings.append(SuspiciousFile(
            path=rel_path,
            reason=f"Hash-like filename {change_type} (possible dropper)",
            detail=rel_path,
        ))


def save_diff_report(diff: SnapshotDiff, output_path: str) -> None:
    """Save snapshot diff to JSON for REDACTS reporting."""
    data = {
        "added": diff.added,
        "removed": diff.removed,
        "modified": diff.modified,
        "unchanged": diff.unchanged,
        "suspiciousFiles": [
            {"path": s.path, "reason": s.reason, "detail": s.detail}
            for s in diff.suspicious_files
        ],
    }
    Path(output_path).write_text(json.dumps(data, indent=2), encoding="utf-8")
